Move classical variables and stretches to CircuitData

eliarbel opened the pull request:
### Summary
This PR moves the storage of variables and stretches from `QuantumCircuit` to `CircuitData` in Rust. 

Closes #14299 

### Details and comments
In this PR, variables and stretches are stored in `CircuitData` using the same approach used in `DAGCircuit`, namely using `ObjectRegistry` for storing the actual `Var` objects (and their unique indices) and `IndexMap` keyed by strings to keep track of identifier names and map them to their corresponding var types. 

Currently `QuantumCircuit` contains logic for ensuring that new vars added to the circuit comply with certain scoping and shadowing rules. This logic has been moved to Rust as well to make `CircuitData` complete w.r.t variables consistency, i.e. it can safe-guard itself when working with it completely in Rust-space.

This PR also allows `converters.rs` to access variables directly rather than via Python calls in Rust.

eliarbel committed:
Add groundwork for vars storage in CircuitData

Adding basic support for vars and identifier storage and access, using
the same approach implemented in DAGCircuit with `ObjectRegsitry` and
various var-types structs. In addition updating `QuantumCircuit` to use
`CircuitData` for input vars.

eliarbel committed:
Handle captured and declared vars in QuantumCircuit

eliarbel committed:
Move stretch to CircuitData

eliarbel committed:
Handle vars & stretches in Rust in dag_to_circuit

eliarbel committed:
Merge remote-tracking branch 'upstream/main' into vars-in-circuitdata

eliarbel committed:
Add pickle/unpickle support + minor misc. stuff

eliarbel committed:
Use CircuitData'a vars/stretches natively in QuantumCircuitData

eliarbel committed:
Clone interners in CircuitData::copy_empty_like

eliarbel committed:
Add scoping rule checks to CircuitData

This commit adds checks to CircuitData::add_var and CircuitData::add_stretch
functions to enforce variable scoping rules. It also includes minor refactoring
and miscellaneous cleanups.

eliarbel committed:
Update BlueprintCircuit & twirling and cleanup

eliarbel committed:
Merge remote-tracking branch 'upstream/main' into vars-in-circuitdata

eliarbel committed:
Add unit tests. Some refactoring and cleanups

eliarbel committed:
Minor touches

eliarbel committed:
Lint and comment

eliarbel committed:
Fix metadata handling following changes in high_level_synthesis.rs

Since this PR no longer uses a custom function to convert a circuit to DAG in
run_high_level_synthesis, we should use DAGCircuit::from_circuit
with QuantumCircuitData initialized with both name and metadata from
the original DAG.

jakelishman commented:
I just came to look at this PR because I have a bug in Rust-space Sabre, which I caught and fixed on `main`, about nested `Var`s in control-flow scopes, and it tracks back to me mistakenly using `DAGCircuit::from_circuit_data` on a block that may have `Var`s; I need this PR merged to avoid having to manually copy over the `Var`s.

Looking through the diff now, I see that I've inadvertently duplicated some work you did about the `VarsCopyMode` thing in https://github.com/Qiskit/qiskit/pull/14557 - I can close that PR if this one merge first.  There'll be merge conflicts whether this PR or https://github.com/Qiskit/qiskit/pull/14496 merges first, but shouldn't be too crazy.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -52,6 +53,9 @@ type CircuitDataState<'py> = (
     Vec<ClassicalRegister>,
     Bound<'py, PyDict>,
     Bound<'py, PyDict>,
+    Vec<(String, PyObject)>,
+    Vec<expr::Var>,
+    Vec<expr::Stretch>,
 );
```
Probably a follow-up: this type alias is getting out of control, and at some point we should probably revisit how we're doing pickling.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -129,11 +133,165 @@ pub struct CircuitData {
     global_phase: Param,
 }
 
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum CircuitVarType {
+    Input = 0,
+    Capture = 1,
+    Declare = 2,
+}
+
+impl From<DAGVarType> for CircuitVarType {
+    fn from(value: DAGVarType) -> Self {
+        match value {
+            DAGVarType::Input => CircuitVarType::Input,
+            DAGVarType::Capture => CircuitVarType::Capture,
+            DAGVarType::Declare => CircuitVarType::Declare,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CircuitVarInfo {
+    var: Var,
+    type_: CircuitVarType,
+}
```
Can we unify the `CircuitVarType` and `DAGVarType` (and friends) structs so there's just one of each?  I think they're identical, and since the sets of representable programs between `CircuitData` and `DAGCircuit` need to overlap, I can't picture a world where these two would diverge.  If they do, I feel like we can handle it when it happens.

eliarbel replied:
I agree in general, see my comment [below](https://github.com/Qiskit/qiskit/pull/14421#issuecomment-2962845227).

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -129,11 +133,165 @@ pub struct CircuitData {
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum CircuitStretchType {
+    Capture = 0,
+    Declare = 1,
+}
+
+impl From<DAGStretchType> for CircuitStretchType {
+    fn from(value: DAGStretchType) -> Self {
+        match value {
+            DAGStretchType::Declare => CircuitStretchType::Declare,
+            DAGStretchType::Capture => CircuitStretchType::Capture,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CircuitStretchInfo {
+    stretch: Stretch,
+    type_: CircuitStretchType,
+}
```
Similarly here: it would be nice to just unify these with the DAG variants, just like how we only have one `PackedInstruction` (and `ObjectRegistry`, etc) that's shared.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -129,11 +133,165 @@ pub struct CircuitData {
+    Alike = 0,
+    Captures = 1,
+    Drop = 2,
+}
```
I've made a DAG version of this `enum` too in https://github.com/Qiskit/qiskit/pull/14557, which would be nice to unify.

eliarbel replied:
Ah, that's nice, I like it better :-) For unifying, how about if we move it to `circuit_data.rs` then? Seems more natural if it's there I think.

eliarbel replied:
Switched to using the `VarsMode` enum in 98a16a87eeaa8156edb0f02f58f08b1fabe33578. 
In addition, I've moved the enum to `circuit`'s `lib.rs` as it is being used by `CircuitData`, `DAGCircuit` and other places.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -164,6 +322,14 @@ impl CircuitData {
+            vars_captured: Vec::new(),
+            vars_declared: Vec::new(),
+            stretches_captured: Vec::new(),
+            stretches_declared: Vec::new(),
```
This is starting to be a lot of fields: I know the data just _is_ complicated, though do you have any ideas if there could be any intermediate structs we could introduce to simplify `CircuitData` itself a bit?  Feels like `vars` are almost invariably going to want to be treated as a block, and a small boiler-plate internal
```rust
struct Vars {
    objects: ObjectRegister<...>,
    inputs: Vec<...>,
    captures: Vec<...>,
    declared: Vec<...>,
}
```
might help with `CircuitData` a bit?  I'm not very sure, it just looks awkward to have so many fields now and there may be some middle-manager bureaucracy that would make it easier to think about.

---

*edit*: From looking at the pickle code, perhaps a low-cost idea would be to use that you already associated integer indices with `VarsType` (i.e. `VarsType::Input = 0`, etc) to have all of `vars_input`, `vars_captured` and `vars_declared` unified into a `[Vec<VarInfo>; 3]`, and just index into that with `VarsType::Input as usize`?  It'd simplify a bunch of `match` statements further down in the pickle code too.

A more boiler-plateish but less "magic" way might be to have
```rust
struct Vars {
    inputs: Vec<VarInfo>,
    captures: Vec<VarInfo>,
    declared: Vec<VarInfo>,
}
impl Index<VarType> for Vars {
    type Item = [VarInfo];
    fn index(&self, type_: VarType) -> &Self::Item {
        match type_ {
            VarType::Input => &self.inputs,
            // ...
        }
    }
}
impl IndexMut<VarType> for Vars {
    // ... same ..
```
and then you'd do `&circuit_data.vars[VarsType::Input]` to get a slice (and the indexer could be dynamically set). It'd all (almost certainly) compile out into simple offset-loads.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -202,12 +368,23 @@ impl CircuitData {
-    pub fn __setstate__(slf: &Bound<CircuitData>, state: CircuitDataState) -> PyResult<()> {
+    pub fn __setstate__(
+        slf: &Bound<CircuitData>,
+        py: Python,
```
You don't need to take the `py` token explicitly - you can just get it for free from `slf` by doing `let py = slf.py()`.

eliarbel replied:
Fixed in 8dd0adf755bd90abbe99d8e16bcbcf6533197377

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -222,6 +399,46 @@ impl CircuitData {
+            IndexMap::<String, CircuitIdentifierInfo>::with_capacity_and_hasher(
+                state.4.len(),
+                RandomState::default(),
+            );
```
Minor style: you should be able to do something like
```rust
borrowed_mut.identifier_info = IndexMap::with_capacity_and_hasher(state.4.len(), Default::default());
```
and type inference will take care of everything.

Fwiw, the `state.4` stuff is why I'm thinking some sort of record-struct for the `state` (or other revisit of pickling) would be could - in review, I can't remember what any of the large tuple indices are, but it'd be clear with a record type.

eliarbel replied:
Right, I'll fix it.

eliarbel replied:
Explicit type annotation removed in 8dd0adf755bd90abbe99d8e16bcbcf6533197377

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -529,29 +746,35 @@ impl CircuitData {
-        res.clbit_indices = self.clbit_indices.clone();
+    ///     CircuitData: The empty copy like self.
+    #[pyo3(name = "copy_empty_like", signature = (*, vars_mode="alike"))]
+    fn py_copy_empty_like(&self, vars_mode: &str) -> PyResult<Self> {
```
If you define a `FromPyObject` for `VarsCopyMode`, you can actually just directly expose `copy_empty_like` and have the "bad string" conversion handling done automatically (disclaimer: that's what I did in #14557).

eliarbel replied:
Done in 98a16a87eeaa8156edb0f02f58f08b1fabe33578

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1045,6 +1268,20 @@ impl CircuitData {
+                {
+                    return Ok(false);
+                }
+            }
```
Two notes:
1. `PyAny::extract` creates new objects with ownership, which is very expensive.  I think you probably meant `downcast`, since you only need immutable borrows onto  `identifier_info`.
2. var-order declarations can be relaxed, but stretch-order declaration is critical to maintain for equality; there's semantics assigned to the order of stretch declarations.

eliarbel replied:
Thanks, I wasn't aware that the order of stretch variables has special semantics. I'll update the logic and will also get rid of the call to `extract`.

eliarbel replied:
Fixed in 8dd0adf755bd90abbe99d8e16bcbcf6533197377

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1317,6 +1805,7 @@ impl CircuitData {
         clbit_indices: BitLocator<ShareableClbit, ClassicalRegister>,
         instructions: I,
         global_phase: Param,
+        variables: Vec<CircuitVar>,
```
Since you take `variables` as necessarily an owned `Vec` of the wrapped `CircuitVar` type, which is not how either `Var` or `Stretch` is used, this API is forcing the caller to re-allocate any vecs they might already have.  At a minimum, we probably should just take `variables: impl Iterator<CircuitVar>` and use the `size_hint` for the capacity estimation, but since this will most likely involve chaining several iterators together to make it (iterator chaining can never be `ExactSizeIterator` because even a chain of two `ExactSizeIterator`s might not be exact size; it might overflow `usize` in length), we probably don't buy anything much from the capacity check.

Do you think the unified wrapper type, which we force callers to construct, is definitely simplifying the API?  (A genuine question - it's not 100% clear to me, but you've used it, and I haven't.)

eliarbel replied:
Point taken about using the `variables` Vec as owned and I agree it's not a perfect solution. My main motivation for introducing it was to enforce variable handling when using `CircuitData::from_packed_instructions`. This function seems to be a complete circuit-to-circuit or dag-to-circuit constructor so I didn't want client code using it to handle variables separately, or even worse, forget to. Ideally, if we had a unified view into `CircuitData` and `DAGCircuit` variables, like `CircuitData::identifer_info` that should suffice. But because of the `CircuitVarType` and `DAGVarType` info systems are separated now I added the unified wrapper `CircuitVar` as a convenience. 

Do you think we should actually enforce variable handling in this `CircuitData::from_packed_instructions`? If so, it goes back to consolidating variable and stretch info handling of `CircuitData` and `DAGCircuit`, in which case once we have it we can just pass whatever container we'll have to `from_packed_instructions` (e.g. `identifier_info`). Alternatively, we can omit the `variables` argument altogether and handle the var and stretch addition in `dag_to_circuit` function which currently is the only place calling `from_packed_instructions`. What do you think?

kevinhartman replied:
If you do plan to keep it @eliarbel, should we rename `CircuitVar` to `CircuitIdenfitier`?

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1957,6 +2451,246 @@ impl CircuitData {
             _ => self.set_global_phase(add_global_phase(&self.global_phase, value)?),
         }
     }
+
```
Fwiw, I got to here in `circuit_data.rs` and then I had to leave for dinner - I haven't looked at anything below here in the PR diff.

jakelishman reviewed (commented):
Thanks for this Eli - I'm keen to get this merged soon.  I started reviewing, but now I have to leave for the day, so here's a partial review.  I noted the point I stopped reading.

eliarbel commented:
Thanks @jakelishman  for the review comments so far. Regarding your suggestion to unify the var and stretch type structs/enums of `CircuitData` and `DAGCircuit`,  I like this approach. However, the two var info systems are not strictly identical currently since `DAGVarInfo` contains 2 fields (for in and out nodes) whereas `CircuitVarType` does not. But this technicality shouldn't really be an obstacle for consolidation. That said, I think it would be better to handle this in a follow-on PR as this seems to increase the scope of this PR too much. Once we have a unified var/stretches info system it would be easier to introduce a struct container to be used in both `CircuitData` and `DAGCircuit` and delegate functionality to it, in a way that would address some of the concerns you raised like serialization, equality checks and data organization in general. What do you think, should we handle this now or defer it to another PR? I'd be happy BTW to handle a follow-on refactor myself.

eliarbel committed:
Address some of Jake's review comments

Relax equality checking to allow variable order independence (modulo the declared stretches).
Minor refactoring.

eliarbel committed:
Merge remote-tracking branch 'upstream/main' into vars-in-circuitdata

eliarbel committed:
Use the `VarsMode` enum in Python's `copy_empty_like`

Following PR #14557, this commit switches to using the `VarsMode` enum
in the Pyo3 function `CircuitData::copy_empty_like`.
Additionaly, since the enum is now used both by `CircuitData` and `DAGCircuit`
and in various placed, it has been moved to the crate level.

eliarbel committed:
Merge remote-tracking branch 'upstream/main' into vars-in-circuitdata

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -129,11 +133,157 @@ pub struct CircuitData {
+    /// Variable identifiers, in order of their addition to the circuit
+    identifier_info: IndexMap<String, CircuitIdentifierInfo>,
+
+    // Var and Stretch indices stored in the circuit
+    vars_input: Vec<Var>,
+    vars_capture: Vec<Var>,
+    vars_declare: Vec<Var>,
+
+    stretches_capture: Vec<Stretch>,
+    stretches_declare: Vec<Stretch>,
```
Not for this PR but why do we store the `Var`s and `Stretch`es in `DAGCircuit` within a `HashSet` as opposed to a `Vec`? As far as I'm aware, if anyone needs to find whether a circuit contains a `Var` the `ObjectRegistry` will already give you the ability to quickly check since they implement `Hash`.

eliarbel replied:
`DAGCircuit::__eq__`  currently depends on the fact that vars and captured stretches are stored in `HashSet`s. It's the only place in `DAGCircuit` that needs `O(1)` lookup of vars. In `CircuitData` the `__eq__` function is implemented differently, using the `identifier_info` field so all the vars and stretch fields can be `Vec`. This has the nice (albeit not strictly required) property that vars and stretch iterators maintain the original insertion order. Note that insertion order of declared stretches has defined semantics so those are stored in `Vec` in `DAGCircuit` as well. 
This divergence will go away once we consolidate the vars and stretches storage.

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1038,13 +1296,76 @@ impl CircuitData {
+                        // in `identifier_info` of the other CircuitData - which match the stretches encountered during the
+                        // iteration here - are monotonically increasing.
+                        if let CircuitStretchType::Declare = rhs_stretch_info.get_type() {
+                            let rhs_stretch_idx =
+                                other_cd.identifier_info.get_index_of(id_name).unwrap();
```
Is there a chance that the `unwrap()` call here might panic? We're getting information based on the `id_name` we extracted from `self_cd` into `other_cd`, how do we know that `id_name` exists in `other_cd` already?

eliarbel replied:
This code is reachable only if `id_name` exists in `other_cd`, as it's guarded by this check at the beginning of the `for` block:
``` rust
let Some(rhs_id_info) = other_cd.identifier_info.get(id_name) else {
    return Ok(false); // Identifier does not exist on the other CircuitData
};
```

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1903,36 +2508,9 @@ impl CircuitData {
-        Ok(empty)
+    /// Returns an iterator over the stored identifiers in order of insertion
+    pub fn identifiers(&self) -> impl ExactSizeIterator<Item = &CircuitIdentifierInfo> {
+        self.identifier_info.iter().map(|info| info.1)
```
I think you should be able to use `IndexMap::values()` for this and still preserve the order.
```suggestion
        self.identifier_info.iter().values()
```

eliarbel replied:
It should actually be `identifier_info.values()`, but yes, you're right it's cleaner this way. 
Changed in d787a0e6fbc251eabbbc0cfc30b92068855fd0aa

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1957,6 +2535,172 @@ impl CircuitData {
+    /// # Returns:
+    ///
+    /// The [Var] index of the variable in the circuit.
+    pub fn add_var(&mut self, var: expr::Var, type_: CircuitVarType) -> PyResult<Var> {
```
Little nitpick there, and might not be for this PR specifically but for rust native functions we should try and stay clear of using `PyResult` for any fallible functions, we should definitely have a Rust native `CircuitError` counter.

eliarbel replied:
I completely agree, however this practice is really quite widespread currently.

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1957,6 +2535,172 @@ impl CircuitData {
+            }
+            CircuitVarType::Capture if !self.vars_input.is_empty() => {
+                return Err(CircuitError::new_err(
+                    "circuits with input variables cannot be enclosed, so cannot be closures",
```
The writing on this error message is a bit strange. I tried adding an article to make it more understandable but it might need to be rewritten further.
```suggestion
                    "circuits with input variables cannot be enclosed, so they cannot be closures",
```

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1206,6 +1527,256 @@ impl CircuitData {
+        Ok(())
+    }
+
+    /// Check if this stretch variable is in the circuit.
+    ///
+    /// Args:
+    ///     var: the variable or name to check.
+    #[pyo3(name = "has_stretch")]
+    fn py_has_stretch(&self, stretch: &Bound<PyAny>) -> PyResult<bool> {
+        if let Ok(name) = stretch.extract::<String>() {
+            Ok(matches!(
+                self.identifier_info.get(&name),
+                Some(CircuitIdentifierInfo::Stretch(_))
+            ))
+        } else {
+            let stretch = stretch.extract::<expr::Stretch>()?;
+            if let Some(CircuitIdentifierInfo::Stretch(info)) =
+                self.identifier_info.get(&stretch.name)
+            {
+                return Ok(&stretch == self.stretches.get(info.stretch).unwrap());
+            }
+            Ok(false)
+        }
+    }
+
+    /// Check if the circuit contains a capture stretch with the given name.
+    #[pyo3(name = "has_captured_stretch")]
+    fn py_has_captured_stretch(&self, name: &str) -> PyResult<bool> {
+        Ok(matches!(
+            self.identifier_info.get(name),
+            Some(CircuitIdentifierInfo::Stretch(stretch_info)) if matches!(stretch_info.type_, CircuitStretchType::Capture)))
+    }
+
+    /// Check if the circuit contains a local stretch with the given name.
+    #[pyo3(name = "has_declared_stretch")]
+    fn py_has_declared_stretch(&self, name: &str) -> PyResult<bool> {
+        Ok(matches!(
+            self.identifier_info.get(name),
+            Some(CircuitIdentifierInfo::Stretch(stretch_info)) if matches!(stretch_info.type_, CircuitStretchType::Declare)))
+    }
```
Many of these methods don't exist or aren't the same in `DAGCircuit` when we end up consolidating both Var's storage in `DAGCircuit` and `CircuitData`, will we make the methods look similar?

eliarbel replied:
First, you're right, only some of the functions are actually used by the classical expression builders so there are differences. Nevertheless, since it was easy and required small amount of additional code, I thought it would be nice to have the Python-facing API of `CircuitData` complete when it comes to working with vars and stretches. That said, I'm OK with removing the functions which are not being used currently if you think it's better this way.
In general, I think consolidating vars storage should have no impact on this aspect if we keep the consolidated struct Rust-only and delegate Python access through `CircuitData` and `DAGCircuit` themselves in a custom way that fits each struct's use-cases.

raynelfss replied:
I'm okay with leaving these in if there is a use for it. My question was more about the latter point you made, which makes sense. Though I understand we may not want to add methods before there is a use for them, I don't think these methods affect anything user-facing.

raynelfss commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1957,6 +2535,172 @@ impl CircuitData {
+    pub fn add_stretch(
+        &mut self,
+        stretch: expr::Stretch,
+        type_: CircuitStretchType,
```
**Nitpick:** I'm not sure if a required argument such as `type_` should have an underscore when it's included as a function argument. Is there a way we could rename it to `stretch_type` or something along those lines?

eliarbel replied:
Changed in d787a0e

raynelfss commented on crates/circuit/src/converters.rs:
```diff
@@ -28,11 +28,6 @@ pub struct QuantumCircuitData<'py> {
     pub data: CircuitData,
     pub name: Option<String>,
```
Are there plans for adding a `name` attribute to `CircuitData`?

raynelfss commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4108,7 +4101,10 @@ def copy(self, name: str | None = None) -> typing.Self:
         """
-        cpy = self.copy_empty_like(name)
+
+        # vars_mode is "drop" since `cpy._data`` is overriden anyway with the call to copy,
```
```suggestion
        # vars_mode is "drop" since ``cpy._data`` is overriden anyway with the call to copy,
```

raynelfss reviewed (commented):
Okay, I've finalized reviewing the rest of the files, I like the direction this is going. I just had a couple more comments.

eliarbel committed:
Update crates/circuit/src/circuit_data.rs

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

eliarbel committed:
Update qiskit/circuit/quantumcircuit.py

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

eliarbel committed:
Address Ray's review comments

eliarbel committed:
Merge branch 'vars-in-circuitdata' of github.com:eliarbel/qiskit into vars-in-circuitdata

eliarbel committed:
Revert change in an exception message

eliarbel committed:
Undo revert of error message change and apply everywhere

eliarbel committed:
Fix a typo

raynelfss reviewed (dismissed):
I think this is good as is. Not sure if @jakelishman had any more comments.

eliarbel committed:
Merge main and resolve conflicts
